design/clean_artifacts_in_csi_flow.md

```diff
@@ -0,0 +1,487 @@
+	}
+
+	p.log.Info("Starting VolumeSnapshotContentRestoreItemAction")
+
+	var snapCont snapshotv1api.VolumeSnapshotContent
+	if err := runtime.DefaultUnstructuredConverter.FromUnstructured(
+		input.Item.UnstructuredContent(), &snapCont); err != nil {
+		return &velero.RestoreItemActionExecuteOutput{},
+			errors.Wrapf(err, "failed to convert input.Item from unstructured")
+	}
+
+	var vscFromBackup snapshotv1api.VolumeSnapshotContent
+	if err := runtime.DefaultUnstructuredConverter.FromUnstructured(
+		input.ItemFromBackup.UnstructuredContent(), &vscFromBackup); err != nil {
+		return &velero.RestoreItemActionExecuteOutput{},
+			errors.Errorf(err.Error(), "failed to convert input.ItemFromBackup from unstructured")
+	}
+
+	// If cross-namespace restore is configured, change the namespace
+	// for VolumeSnapshot object to be restored
+	newNamespace, ok := input.Restore.Spec.NamespaceMapping[snapCont.Spec.VolumeSnapshotRef.Namespace]
+	if ok {
+		// Update the referenced VS namespace to the mapping one.
+		snapCont.Spec.VolumeSnapshotRef.Namespace = newNamespace
+	}
+
+	// Reset the ResourceVersion and UID of referenced VolumeSnapshot.
+	snapCont.Spec.VolumeSnapshotRef.ResourceVersion = ""
+	snapCont.Spec.VolumeSnapshotRef.UID = ""
+
+	// Set the DeletionPolicy to Retain to avoid VS deletion will not trigger snapshot deletion
+	snapCont.Spec.DeletionPolicy = snapshotv1api.VolumeSnapshotContentRetain
+
+	if vscFromBackup.Status != nil && vscFromBackup.Status.SnapshotHandle != nil {
+		snapCont.Spec.Source.VolumeHandle = nil
+		snapCont.Spec.Source.SnapshotHandle = vscFromBackup.Status.SnapshotHandle
+	} else {
+		p.log.Errorf("fail to get snapshot handle from VSC %s status", snapCont.Name)
+		return nil, errors.Errorf("fail to get snapshot handle from VSC %s status", snapCont.Name)
+	}
+
+	additionalItems := []velero.ResourceIdentifier{}
+	if csi.IsVolumeSnapshotContentHasDeleteSecret(&snapCont) {
+		additionalItems = append(additionalItems,
+			velero.ResourceIdentifier{
+				GroupResource: schema.GroupResource{Group: "", Resource: "secrets"},
+				Name:          snapCont.Annotations[velerov1api.PrefixedSecretNameAnnotation],
+				Namespace:     snapCont.Annotations[velerov1api.PrefixedSecretNamespaceAnnotation],
+			},
+		)
+	}
+
+	vscMap, err := runtime.DefaultUnstructuredConverter.ToUnstructured(&snapCont)
+	if err != nil {
+		return nil, errors.WithStack(err)
+	}
+
+	p.log.Infof("Returning from VolumeSnapshotContentRestoreItemAction with %d additionalItems",
+		len(additionalItems))
+	return &velero.RestoreItemActionExecuteOutput{
+		UpdatedItem:     &unstructured.Unstructured{Object: vscMap},
+		AdditionalItems: additionalItems,
+	}, nil
+}
+```
+
+#### Use Async Operation for VolumeSnapshot restore
+The VS and VSC restore returns after creation. By then, the snapshot creation is still in progress. It's possible the snapshot creation fails later.
+
+To avoid neglect in restore, propose to add async operation in the VS RestoreItemAction.
+
```

Comment:
this is a valuable improvement in restore flow. 
However I would request you put this behind a feature flag - such that we can disable this polling and revert to current velero behaviour in case of issues.
So far we have relied on velero being fire and forget for PVC restore, this change might have impact in downstream, hence a flag would help toggle this behaviour to avoid regressions. 
Flag can be true by default.

Reply:
OK. Removed the async retore part.